sklearn/datasets/tests/test_base.py

```diff
@@ -31,97 +33,92 @@
 from sklearn.utils.testing import assert_raises
 
 
-DATA_HOME = tempfile.mkdtemp(prefix="scikit_learn_data_home_test_")
-LOAD_FILES_ROOT = tempfile.mkdtemp(prefix="scikit_learn_load_files_test_")
-TEST_CATEGORY_DIR1 = ""
-TEST_CATEGORY_DIR2 = ""
-
-
 def _remove_dir(path):
     if os.path.isdir(path):
         shutil.rmtree(path)
 
 
-def teardown_module():
-    """Test fixture (clean up) run once after all tests of this module"""
-    for path in [DATA_HOME, LOAD_FILES_ROOT]:
-        _remove_dir(path)
+@pytest.fixture(scope="module")
+def data_home(tmpdir_factory):
+    tmp_file = str(tmpdir_factory.mktemp("scikit_learn_data_home_test"))
+    yield tmp_file
+    _remove_dir(tmp_file)
+
 
+@pytest.fixture(scope="module")
+def load_files_root(tmpdir_factory):
+    tmp_file = str(tmpdir_factory.mktemp("scikit_learn_load_files_test"))
+    yield tmp_file
+    _remove_dir(tmp_file)
 
-def setup_load_files():
-    global TEST_CATEGORY_DIR1
-    global TEST_CATEGORY_DIR2
-    TEST_CATEGORY_DIR1 = tempfile.mkdtemp(dir=LOAD_FILES_ROOT)
-    TEST_CATEGORY_DIR2 = tempfile.mkdtemp(dir=LOAD_FILES_ROOT)
+
+@pytest.fixture
+def test_category_dir_1(load_files_root):
+    TEST_CATEGORY_DIR1 = tempfile.mkdtemp(dir=load_files_root)
     sample_file = tempfile.NamedTemporaryFile(dir=TEST_CATEGORY_DIR1,
                                               delete=False)
     sample_file.write(b("Hello World!\n"))
     sample_file.close()
+    yield str(TEST_CATEGORY_DIR1)
```

Comment:
Yeah, but the returned object is a `py.path.local` see [tmdir defintion](https://github.com/pytest-dev/pytest/blob/075faa5e2b860151a7680cb1d4c50590d3b619cf/src/_pytest/tmpdir.py#L120). [Its readme](https://github.com/pytest-dev/py ) recommends using pathlib instead, and is staged for removal in 4.0 https://github.com/pytest-dev/pytest/issues/1260. All of this to say that I wouldn't expect contributors to know that it has the `.strpath` method, and assuming that it can convert to string path correctly is the weakest assumption I can make. 

Also that's the recommended approach for converting pathlib objects to str (https://www.python.org/dev/peps/pep-0428/#representing), so when the migration happen in pytest4 the code won't need to be changed.

Reply:
Fine with me, I was not aware of this.